pkg/identity/spiffe/principal.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spiffe
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/sigstore/fulcio/pkg/ca/x509ca"
+	"github.com/sigstore/fulcio/pkg/config"
+	"github.com/sigstore/fulcio/pkg/identity"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+)
+
+type principal struct {
+	// spiffe ID
+	id string
+
+	// OIDC issuer url
+	issuer string
+}
+
+func PrincipalFromIDToken(ctx context.Context, token *oidc.IDToken) (identity.Principal, error) {
+	cfg, ok := config.FromContext(ctx).GetIssuer(token.Issuer)
+	if !ok {
+		return nil, errors.New("invalid configuration for OIDC ID Token issuer")
+	}
+
+	if err := validSpiffeID(token.Subject, cfg.SPIFFETrustDomain); err != nil {
+		return nil, err
+	}
+
+	// Now issue cert!
+	return principal{
+		id:     token.Subject,
+		issuer: token.Issuer,
+	}, nil
+
+}
+
+func validSpiffeID(id, trustDomain string) error {
```

Comment:
refactor nit: this is fine because it's small, but this function includes multiple other refactors, moving a function and variable renaming.